Simplify board parsing

aexg opened the pull request.

aexg committed:
Parse row and column counts together to reuse search results

aexg committed:
Parse score and moves together with board size to reuse search results

aexg committed:
Read in the board without isolating it

aexg committed:
Put cursor in the middle if it was missing

Jiwan commented on parser.hpp:
```diff
@@ -97,21 +117,32 @@ CONSTEXPR CandyState decode_candy_state(char up_char)
+constexpr int noncandy_rows = 3;  // title, border, border
 
-constexpr auto parse_board_column_count(const constexpr_string_view& str)
+constexpr auto parse_board_parameters(auto&& str)
```
Since this is a string_view, I actually think that here the best would be to take by value

aexg replied:
Done! Now send by value

Jiwan commented on parser.hpp:
```diff
@@ -97,21 +117,32 @@ CONSTEXPR CandyState decode_candy_state(char up_char)
+    return ret;
 }
 
 constexpr auto isolate_board_game(auto&& str)
```
I am guessing that this function has no use anymore. We could remove it.

aexg replied:
Removed :-)

Jiwan commented on parser.hpp:
```diff
@@ -13,6 +13,26 @@
+    int moves;
+};
+
+constexpr int extract_int(auto it)
```
As a security it would be nice to take begin and end here (even though this would result in a compile-time crash if we go too far in the buffer :D).

aexg replied:
Added limit for number of read chars

Jiwan commented on parser.hpp:
```diff
@@ -145,10 +176,11 @@ constexpr auto isolate_board_game(auto&& str)
 
-    constexpr int column_count = parse_board_column_count(board_string);
-    constexpr int row_count = parse_board_row_count(board_string);
+    constexpr auto board_parameters = parse_board_parameters(get_game_state_string());
```
Nice, only one pass to get many parameters at once :)

Jiwan commented on parser.hpp:
```diff
@@ -97,21 +117,32 @@ CONSTEXPR CandyState decode_candy_state(char up_char)
-    int row_count = count(str.cbegin(), str.cend(), '\n');
+    auto gs_begin = str.cbegin();
+    auto gs_end = str.cend();
+    auto lefttop_corner = find(gs_begin + 1, gs_end, '\n');
```
Ins't the lefttop_corner actually one offset after '\n'?

aexg replied:
True, renamed variables for "anchors"!

Jiwan commented on parser.hpp:
```diff
@@ -158,39 +190,46 @@ CONSTEXPR auto parse_board(GameString&& get_game_state_string)
             };
 
-            if (board_string[candy_state_selected_or_hover] == '[') {
+            char selected_or_hovered = board_string[candy_type_index - 1];
```
Nice!

Jiwan commented on parser.hpp:
```diff
@@ -220,18 +259,15 @@ template <class GameString>
-    CONSTEXPR auto moves = parse_moves(get_game_state_string);
 
-    return std::make_tuple(board, score, moves);
+    return board;
```
In retrospective, I like structs a lot better.

aexg replied:
Made game state as struct in a separate commit since it affected several files (still same PR), also removed parser include from engine file since engine does not use parser

Jiwan reviewed (commented):
Maybe so small issues should be addressed, apart from that really great improvement!

aexg committed:
Adjust parsing according to review comments

- add limit when reading integer from string
- send string_view by value to function parse_board_parameters
- rename top_corner anchors names to more relevant
- remove non-used function isolate_board_game

aexg committed:
Store game state in a struct instead for tuple

aexg commented:
You are right about small issues. I strived to keep as much code as possible unaffected for more readable diffs and more clear purpose of changes. Now it is time for me to clean up afterwards!

Jiwan reviewed (approved):
Looks good!
